Add mesh support

MoritzWM opened the pull request:
Hello,
this is not a finished pull request, I just wanted to gather some feedback on the mesh and general data parsers. I developed [napari-imodmodel](https://github.com/MoritzWM/napari-imodmodel) alongside it and the meshes are displayed correctly.

I added a meshes list to Object. Each Mesh is a list of vertices, indices and extra data.
The vertices are already reshaped to (-1, 3) during parsing, the indices are reshaped on-demand because the original list is required to calculate the indices for the extra data. I find this a bit inconsistent and wanted to ask if you have a better idea.

Only one type of mesh is supported right now (the one with control int -25), I don't have a mesh at hand where they are stored differently.

I didn't yet add any tests.

Please let me know what you think and how you would like to have it implemented.

alisterburt commented on src/imodmodel/parsers.py:
```diff
@@ -56,12 +73,8 @@ def _parse_object_header(file: BinaryIO) -> ObjectHeader:
-    for _ in range(header.contsize):
-        _parse_control_sequence(file)
-        contours.append(_parse_contour(file))
-    return Object(contours=contours)
+    _parse_object_header(file)
+    return Object()
```
am I reading this right that currently you are not returning a valid object here? i.e. this breaks current parsing

MoritzWM replied:
Yes, I changed the code such that the object doesn't "parse itself" anymore. I found it easier to determine the next chunk ID in parse_model() and call the appropriate parsing function there. Otherwise, if _parse_object() runs into the end of its description in the file, it has already consumed the next chunk ID (let's say it's another object). When it returns to the caller, the caller doesn't know what the next chunk ID is. I was at first playing with moving the file pointer back by 4 bytes, but that felt hacky and I'm not sure if this works everywhere.

 _parse_object_header and _parse_object are pretty much equivalent right now, and the returned Object() is empty which I now realize is not so nice. My solution would be to merge _parse_object() and _parse_object_header() and also merge Object and ObjectHeader.

Before, you were using header.contsize to parse the appropriate number of contours. I would rather just parse every contour, mesh and whatever else there is and then, when the parsing is done, validate that header.contsize agrees with the actual number of contours and throw a warning otherwise.

alisterburt replied:
This sounds cleaner and I'm open to unification/cleanup, thanks for taking the time to engage with my code - that work can be in a separate PR before or after this one if you want to keep this simpler 🙂

alisterburt commented on src/imodmodel/models.py:
```diff
@@ -88,11 +96,49 @@ class Contour(BaseModel):
+    class Config:
+        arbitrary_types_allowed = True
+
+    def reshaped_indices(self) -> np.ndarray:
+        return self.indices[np.where(self.indices >= 0)].reshape((-1, 3))
```
model looks good here - rather than having a reshaped indices method (or a [property](https://www.tutorialsteacher.com/python/property-decorator), for cleaner syntax) I would probably implement a [`field_validator`](https://docs.pydantic.dev/latest/usage/validators/#field-validators) to instead ensure the field has the correct shape

MoritzWM replied:
Yes a property is a very good idea. I could make the raw indices (and maybe also vertices) private variables and put the cleanup and reshaping into properties.

I've never used pydantic before, but that sounds like a plan.

alisterburt replied:
A field validator is a little more appropriate than a property here, it's basically where you put logic to ensure data in a field is parsed correctly (validated) so you can just use the field rather than adding another property - I vote let's go with a validator here 🙂

MoritzWM replied:
I think a validator is a good idea, but doesn't solve the problem why I wrote the reshaped_indices method:
The indices in the file look like this: `[-25, 1, 2, 3, 4, .... -22, -1]`. All values >= 0 are actual indices, all values < 0 are special values:
```
  -1   end of list array                
  -20  next item on list is normal vector.
  -21  begin polygon with vertices and vertex indices only <--- what I'm doing right now and what Napari wants
  -22  end polygon
  -23  begin vertex,normal polygon pairs with normal,vertex indices
  -24  begin large convex polygon with normals
  -25  begin vertex,normal polygon pairs with vertex indices <--- the only type of mesh that I implemented so far
```
To get a "clean" indices list for Napari, I'm removing all values < 0 and reshape it to (-1, 3), so I'm storing vertex indices only. However, to correctly interpret a MEST chunk one needs the raw indices list with control values, that's why I'm storing it like that in the object and only create the "clean" version on demand.

Maybe the better way would be:
- validate the indices during parsing: check the shape, check that all mesh types are supported and so on
- store the validated list as-is
- write a function/property that produces the indices in the shape and form that one needs
  - vertex_indices
  - raw_indices
  - ...?

alisterburt replied:
wow... this format is a mess 😆 thank you for patiently explaining - in this case what you suggest makes perfect sense - store the data in the file then further parse somewhere downstream, a property for the simple case seems very reasonable

alisterburt commented on src/imodmodel/parsers.py:
```diff
@@ -40,6 +43,20 @@ def _parse_from_format_str(file: BinaryIO, format_str: str) -> Tuple[Any, ...]:
     return struct.unpack(file.read(struct.size))
 
 
+def _parse_from_type_flags(file: BinaryIO, flags: int) -> Any:
+    flag_mask, flag_int, flag_float, flag_short, flag_byte = 3, 0, 1, 2, 3
+    if flags & flag_mask == flag_int:
+        return _parse_from_format_str(file, '>i')[0]
+    elif flags & flag_mask == flag_float:
+        return _parse_from_format_str(file, '>f')[0]
+    elif flags & flag_mask == flag_short:
+        return _parse_from_format_str(file, '>2h')
+    elif flags & flag_mask == flag_byte:
+        return _parse_from_format_str(file, '>4c')
+    else:
+        raise ValueError(f'Invalid flags: {flags}')
```
this looks like witchcraft, could you explain? 🙂

MoritzWM replied:
I've added some explanation to the code, but basically I'm checking for a 2-bit flag for the type of `index` and `value`, which are stored as unions in the file. The first two bits indicate the type of `index`, the following two the type of `value`. That's why I'm shifting the flags by two bits in [_parse_general_storage](https://github.com/MoritzWM/imodmodel/blob/26429c3cd1a6b5715b92a04d3cd0044719e7d794/src/imodmodel/parsers.py#L133).

There are 4 more bit flags described in the model spec:
```
Bit 4 on : "index" is not really an index
Bit 5 on : Revert to default value
Bit 6 on : Index is surface number, not contour
Bit 7 on : The change applies to only one point
```
But for parsing index and value, they are not directly relevant I suppose.

alisterburt replied:
Cool, thank you! Would you mind adding tests covering this? It's a bit complicated and I want to make sure I don't break it in the future

MoritzWM replied:
Done! I added tests for [_parse_general_storage](https://github.com/MoritzWM/imodmodel/blob/dad7854104d11826c6e334c3246107b9ad6d6e67/tests/test_parsers.py#L217) and for [_parse_from_type_flags](https://github.com/MoritzWM/imodmodel/blob/dad7854104d11826c6e334c3246107b9ad6d6e67/tests/test_parsers.py#L208), and immediately discovered a small bug.

alisterburt reviewed (commented):
Wow @MoritzWM - amazing PR! I have a couple of questions below but it's a great start, love where this is headed and would love to properly support models containing meshes 🙂 

Feel free to ping if you need a hand

alisterburt commented:
silly question - I output a simple dataframe from the read function in the case of contours, it feels like it won't be possible to maintain this once mesh support is added

potential routes:
- more complicated files should go via the parser and extract what they want for the object, leave the `imodmodel.read()` API for simple model files containing points in multiple contours?
- add different return types for meshes in the `read` API

I'm not sure what's best here, have you thought about this?

alisterburt commented on src/imodmodel/models.py:
```diff
@@ -11,6 +11,14 @@ class ID(BaseModel):
+    """https://bio3d.colorado.edu/imod/doc/binspec.html"""
+    type: int
+    flags: int
+    index: Any
+    value: Any
```
is it possible to type these more strictly? Maybe as a `Union[bla, bla]`?

MoritzWM replied:
I had them stored as Union[int, float, Tuple[int, int], Tuple[int, int, int, int]] before but then pydantic was converting floats to int or vice versa. I guess I'd have to read the pydantic docs a bit more to understand why, but I agree this should be typed better.

alisterburt replied:
Yep automatic type coercion has some kind of order implied in the type definition - you can get around this with custom logic in a validator 🙂

MoritzWM replied:
Hey,
just got the time to work on this again.
I implemented it using smart_union=True, but this has been removed in pydantic 2. I didn't notice this before because I have Napari installed in the same venv, which requires pydantic<2. 

Maybe pydantic<2 should be explicitly required.

alisterburt replied:
Agh this whole pydantic updgrade has been a bit of a nightmare - I'm fine with requiring <2 for now since your desired use case is to coexist with napari ☺️

MoritzWM replied:
OK I'll add pydantic<2 as a dependency

alisterburt commented on src/imodmodel/models.py:
```diff
@@ -88,11 +96,49 @@ class Contour(BaseModel):
+class Mesh(BaseModel):
+    """https://bio3d.colorado.edu/imod/doc/binspec.html"""
+    header: MeshHeader
+    vertices: np.ndarray # vert
+    indices: np.ndarray # list
```
could you please remove these comments and add description/link to description of IMOD index list to docstring before merge

alisterburt commented on src/imodmodel/models.py:
```diff
@@ -88,11 +96,49 @@ class Contour(BaseModel):
+    def reshaped_indices(self) -> np.ndarray:
+        return self.indices[np.where(self.indices >= 0)].reshape((-1, 3))
+
+    def extra_values(self) -> np.ndarray:
+        # The extra values are index, value  pairs
+        # However, the index is an index into the indices array,
+        # not directly an index of a vertex.
+        # Furthermore, the index has to be fixed because
+        # the original indices array has special command values (-25, -22, -1, ...)
+        values = np.zeros((len(self.vertices),))
+        for extra in self.extra:
+            if not (extra.type == 10 and isinstance(extra.index, int)):
+                continue
+            # removed_indices = np.nonzero(self.indices < 0)[0]
+            # index_fixed = extra.index - len(np.nonzero(removed_indices < extra.index)[0])
+            values[self.indices[extra.index]] = extra.value
+        return values
```
sorry for another silly question - lack of experience, these are features stored on faces of the mesh?

MoritzWM replied:
I have renamed the function to face_values and turned it into a property.

alisterburt commented on src/imodmodel/models.py:
```diff
@@ -88,11 +96,49 @@ class Contour(BaseModel):
+        for extra in self.extra:
+            if not (extra.type == 10 and isinstance(extra.index, int)):
+                continue
+            # removed_indices = np.nonzero(self.indices < 0)[0]
+            # index_fixed = extra.index - len(np.nonzero(removed_indices < extra.index)[0])
```
remove before merge

alisterburt commented on src/imodmodel/parsers.py:
```diff
@@ -91,6 +130,19 @@ def _parse_imat(file: BinaryIO) -> IMAT:
+    if size % 12 != 0:
+        raise ValueError(f"Chunk size not divisible by 12: {size}")
+    storages = list()
+    for _ in range(size // 12):
```
```suggestion
    n_chunks = size // 12
    for _ in range(n_chunks):
```

alisterburt reviewed (commented):
this is getting really close! few comments below 

do you see anything still missing prior to merge?

MoritzWM commented:
> silly question - I output a simple dataframe from the read function in the case of contours, it feels like it won't be possible to maintain this once mesh support is added
> 
> potential routes:
> 
>     * more complicated files should go via the parser and extract what they want for the object, leave the `imodmodel.read()` API for simple model files containing points in multiple contours?
> 
>     * add different return types for meshes in the `read` API
> 
> 
> I'm not sure what's best here, have you thought about this?

Not really to be honest. I personally think it would be easiest to return an ImodModel object. If I remember right it's how it's also done for in the mrcfile library. This object could also be used for writing models in the future (if that should ever be  implemented).

alisterburt commented:
Sounds good - to maintain compatibility with the previous API then this should be accessed via the ImodModel object with a .from_file(path: PathLike) class method (might already be the case, haven’t looked :))Sent from mobile - apologies for brevityOn 9 Aug 2023, at 19:06, Moritz Wachsmuth-Melm ***@***.***> wrote:﻿

silly question - I output a simple dataframe from the read function in the case of contours, it feels like it won't be possible to maintain this once mesh support is added
potential routes:
* more complicated files should go via the parser and extract what they want for the object, leave the `imodmodel.read()` API for simple model files containing points in multiple contours?

* add different return types for meshes in the `read` API

I'm not sure what's best here, have you thought about this?

Not really to be honest. I personally think it would be easiest to return an ImodModel object. If I remember right it's how it's also done for in the mrcfile library. This object could also be used for writing models in the future (if that should ever be  implemented).

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

MoritzWM commented:
Hi,
sorry for taking so long. I've added an example to the readme. The from_file function was already there, so I just added ImodModel to __all__.

In the future, one could add a nicer API for contour extra values, but I guess that's stuff for a future pull request.

Let me know if there's anything you're still missing in the PR.

jojoelfe commented:
Hi,

I've been thinking about importing imod meshes into blender so this PR would really help with that. I am just checking what the status is and whether I can maybe help?

Best,

Johannes

alisterburt commented:
Hey @jojoelfe and @MoritzWM ! @MoritzWM sorry for the huge delay here, this came in whilst I was moving to the US and got lost in the fold. @jojoelfe I've added you as a maintainer to the project - please feel free to merge/work on this PR with @MoritzWM and you can cut a release whenever you feel ready! pushing to main with the tag "v<ver>" where `<ver>` is the semantic version  should trigger the test/release process on github actions :-) 

If anything isn't clear/any questions pop up I'm here and happy to help!

jojoelfe commented:
Thanks @alisterburt ! @MoritzWM Are you still interested in joining this? I haven't looked into it in great detail, going over the comments briefly I only wonder if this still should be using pydantic 1 or make the switch to 2?

alisterburt commented:
@jojoelfe I'm keen for pydantic 2, I assume the ecosystem won't take too long to fully move over and I'd rather be ahead than behind. If compatiblity can be maintained for pydantic 1 that would be great but no hard requirement :-)

alisterburt commented:
@MoritzWM I've also made you a maintainer here! Please still open PRs with changes and try to keep changes minimal to simplify review - review should still be requested but this should stop you from getting blocked if I or @jojoelfe is unavailable and you need to move forwards :-)

jojoelfe commented:
I haven't tested the import into blender yet, but opening some of my files works without issue. To maintain compatibility with both pydantic v1 and v2 we can just:

```python
from pydantic.version import VERSION as PYDANTIC_VERSION
```

and then make the config class conditional:

```python
class GeneralStorage(BaseModel):
    """https://bio3d.colorado.edu/imod/doc/binspec.html"""
    type: int
    flags: int
    index: Union[float, int, Tuple[int, int], Tuple[int, int, int, int]]
    value: Union[float, int, Tuple[int, int], Tuple[int, int, int, int]]
  
    if PYDANTIC_VERSION < '2.0':
        class Config:
            smart_union = True
```

smart_union is only deprecated in v2 because it is the default behavior.

This seems a small price to pay for making sure this works with most packages. We should make a hatch test matrix with both pydantic versions, but that can be a separate PR.

jojoelfe commented:
I tested this with blender and it works great. I pushed the changes I described above and enabled tests against both pydantic versions.

@alisterburt @MoritzWM as long as there are no vetos, I'll go ahead and merge

alisterburt commented:
Amazing work both, thanks you!!

alisterburt commented:
I pushed v0.0.9 which includes this, should be up on pypi once the test/deploy matrix finishes
https://github.com/teamtomo/imodmodel/actions/runs/7500839757

Thanks both! Would be great if we could follow up with documenting this :-)
